### PR TITLE
Improve Dockerfile

### PR DESCRIPTION
**Pull request by Luzifer**
* Added VOLUME to enable users to keep a state of the repo outside the container
* Added ipfs user to let ipfs run as a normal user and not as root
* Set IPFS_PATH to push IPFS to use the exposed dir
* Improved start script to be more verbose about errors

### CODE REVIEW

**Comment by jbenet** [comment]
@Luzifer said 
> <luzifer> jbenet: that is the current state: https://github.com/ipfs/go-ipfs/pull/1128/files
> <luzifer> (do not merge, its broken)

**Comment by Luzifer** [comment]
```
> docker run -ti b71e52a36079
initializing ipfs node at /data/ipfs
generating 2048-bit RSA keypair...done
peer identity: QmS4VcerJH8hCeeQUpXFbMtDaT6b2dDqGPG9hQUs6NHjme
Error: no ipfs repo found. please run: ipfs init
Error: no ipfs repo found. please run: ipfs init
Initializing daemon...
Error: no ipfs repo found. please run: ipfs init
```

???!???!

**Comment by wking** [comment]
useradd and the later VOLUME and USER commands are not expensive
operations, but shifting them before the cache-busting ADD will let
you share the layers between images that have different ADD-ed
content:

  …
  * ‘FROM golang:1.4’ layer
  …
  * ‘RUN useradd …’ layer
  * ‘USER …’ layer
  * ‘VOLUME …’ layer
  |\
  | * ‘ADD . …’ layer pulling in the current master
  | …
  * ‘ADD . …’ layer pulling in some other commit
  …

**Comment by Luzifer** [comment]
Good point. (even though dockerhub is regenerating the build as a whole instead of reusing)

**Comment by Luzifer** [comment]
```
luzifer@knut-workstation01-2 ~> docker run -ti -v /tmp/ipfs:/data/ipfs fe8290c828e7
ipfs version 0.3.2
initializing ipfs node at /data/ipfs
generating 2048-bit RSA keypair...done
peer identity: QmRBLJ76A24EymKQSCXU91fBAFyocuAbFCMJmh4373Lrxx
to get started, enter:

	ipfs cat /ipfs/Qmcqtw8FfrVSBaRmbWwHxt3AuySBhJLcvmFYi3Lbc4xnwj/readme

Initializing daemon...
API server listening on /ip4/0.0.0.0/tcp/5001
Gateway (readonly) server listening on /ip4/0.0.0.0/tcp/8080
^C
Received interrupt signal, shutting down...
Publishing!
21:12:31.400 ERROR     ipnsfs: leveldb: closed system.go:79
luzifer@knut-workstation01-2 ~> docker run -ti -v /tmp/ipfs:/data/ipfs 78a50264f5fc
Running ipfs version 0.3.2...
Found ipfs repository. Not initializing.
Initializing daemon...
API server listening on /ip4/0.0.0.0/tcp/5001
Gateway (readonly) server listening on /ip4/0.0.0.0/tcp/8080
^C
Received interrupt signal, shutting down...
Publishing!
21:14:31.656 ERROR     ipnsfs: leveldb: closed system.go:79
luzifer@knut-workstation01-2 ~>
```

**Comment by Luzifer** [comment]
@wking / @jbenet PTAL

**Please let @jbenet merge this because of this:**

```
22:45:48 <@jbenet> if the dockerfile works as is, i'd prefer not changing it before tonight
                   (http://meetup.com/EthereumSiliconValley/events/221350594/)
```

File: Dockerfile [hunk 1]
```diff
@@ -1,14 +1,23 @@
...
+
 ADD . /go/src/github.com/ipfs/go-ipfs
 RUN cd /go/src/github.com/ipfs/go-ipfs/cmd/ipfs && go install
 
 RUN cp /go/src/github.com/ipfs/go-ipfs/bin/container_daemon /usr/local/bin/start_ipfs && \
     chmod 755 /usr/local/bin/start_ipfs
 
-EXPOSE 4001 5001 8080
-# 4001 = Swarm, 5001 = API, 8080 = HTTP transport
+USER ipfs
```
Comment (jbenet) [inline Dockerfile]: from @wking's comment, i think this line should be lifted above the `ADD` ?

**Comment by jbenet** [comment]
thanks for taking a look @wking ! :+1: -- any of your thoughts here are way better than mine.

I'll defer to @wking's opinions here. Let's merge when he thinks it's ready.

**Comment by jbenet** [comment]
thanks for the PR @Luzifer

File: Dockerfile (hunk 1 above)
Comment (Luzifer) [inline Dockerfile]: As the user `ipfs` does not own the `/go/src` and `/go/bin` dirs its not useful to change the user earlier than this.

Comment (wking) [inline Dockerfile]: On Mon, Apr 27, 2015 at 08:09:33AM -0700, Knut Ahlers wrote:
> >  ADD . /go/src/github.com/ipfs/go-ipfs
> >  RUN cd /go/src/github.com/ipfs/go-ipfs/cmd/ipfs && go install
> >  
> >  RUN cp /go/src/github.com/ipfs/go-ipfs/bin/container_daemon /usr/local/bin/start_ipfs && \
> >      chmod 755 /usr/local/bin/start_ipfs
> >  
> > -EXPOSE 4001 5001 8080
> > -# 4001 = Swarm, 5001 = API, 8080 = HTTP transport
> > +USER ipfs
> 
> As the user `ipfs` does not own the `/go/src` and `/go/bin` dirs its
> not useful to change the user earlier than this.

Good point (the ‘ipfs’ user can't add to /go/src/…).  This PR looks
good to me as it stands.  The Travis failure is one of the sporadic
“No output has been received in the last ... minutes”, so I wouldn't
worry about that.  +1 to merge.

**Pull request merged**